File: Server/src/alerts/alertService.ts
import type { AppConfig } from '../app/config.js'

const ALERT_HTTP_TIMEOUT_MS = 10_000

export type AlertSeverity = 'info' | 'warning' | 'critical'

export type AlertEvent = {
  key: string
  severity: AlertSeverity
  title: string
  message: string
  metadata?: Record<string, unknown>
}

type AlertPayload = {
  key: string
  severity: AlertSeverity
  title: string
  message: string
  metadata: Record<string, unknown>
  occurredAt: string
  source: string
}

const severityPrefix: Record<AlertSeverity, string> = {
  info: 'INFO',
  warning: 'WARNING',
  critical: 'CRITICAL',
}

const compact = (value: string) => value.trim()

export class AlertService {
  private readonly lastSentByKey = new Map<string, number>()

  constructor(private readonly config: AppConfig) {}

  async notify(event: AlertEvent) {
    if (!this.config.ALERTS_ENABLED) {
      return
    }

    const now = Date.now()
    const previousSentAt = this.lastSentByKey.get(event.key) ?? 0
    if (now - previousSentAt < this.config.ALERT_COOLDOWN_MS) {
      return
    }

    this.lastSentByKey.set(event.key, now)

    const payload = this.buildPayload(event, now)
    const tasks: Array<Promise<void>> = []

    if (this.config.ALERT_SLACK_WEBHOOK_URL) {
      tasks.push(this.sendSlack(payload))
    }

    if (this.hasTwilioChannel()) {
      tasks.push(this.sendTwilioWhatsapp(payload))
    }

    if (this.config.ALERT_EMAIL_WEBHOOK_URL) {
      tasks.push(this.sendEmailWebhook(payload))
    }

    if (!tasks.length) {
      return
    }

    const results = await Promise.allSettled(tasks)
    const failures = results.filter((result) => result.status === 'rejected')

    if (!failures.length) {
      return
    }

    failures.forEach((failure) => {
      if (failure.status === 'rejected') {
        console.error('[alert-service] alert delivery failed', failure.reason)
      }
    })

    // Allow immediate retry on the next trigger if every channel failed.
    if (failures.length === results.length) {
      this.lastSentByKey.delete(event.key)
    }
  }

  private buildPayload(event: AlertEvent, now: number): AlertPayload {
    return {
      key: event.key,
      severity: event.severity,
      title: compact(event.title),
      message: compact(event.message),
      metadata: event.metadata ?? {},
      occurredAt: new Date(now).toISOString(),
      source: 'derton-finance-server',
    }
  }

  private hasTwilioChannel() {
    return Boolean(
      this.config.ALERT_WHATSAPP_TWILIO_ACCOUNT_SID.trim() &&
        this.config.ALERT_WHATSAPP_TWILIO_AUTH_TOKEN.trim() &&
        this.config.ALERT_WHATSAPP_TWILIO_FROM.trim() &&
        this.config.ALERT_WHATSAPP_TWILIO_TO.trim(),
    )
  }

  private async sendSlack(payload: AlertPayload) {
    const metadataRows = Object.entries(payload.metadata)
      .map(([key, value]) => `${key}: ${String(value)}`)
      .join('\n')

    const text = [
      `[${severityPrefix[payload.severity]}] ${payload.title}`,
      payload.message,
      `Source: ${payload.source}`,
      `Time: ${payload.occurredAt}`,
      metadataRows ? `Details:\n${metadataRows}` : '',
    ]
      .filter(Boolean)
      .join('\n')

    await this.postJson(this.config.ALERT_SLACK_WEBHOOK_URL, {
      text,
    })
  }

  private async sendTwilioWhatsapp(payload: AlertPayload) {
    const sid = this.config.ALERT_WHATSAPP_TWILIO_ACCOUNT_SID.trim()
    const token = this.config.ALERT_WHATSAPP_TWILIO_AUTH_TOKEN.trim()
    const from = this.withWhatsappPrefix(this.config.ALERT_WHATSAPP_TWILIO_FROM)
    const to = this.withWhatsappPrefix(this.config.ALERT_WHATSAPP_TWILIO_TO)
    const endpoint = `https://api.twilio.com/2010-04-01/Accounts/${encodeURIComponent(sid)}/Messages.json`
    const body = new URLSearchParams({
      From: from,
      To: to,
      Body: this.toPlainText(payload),
    })

    const auth = Buffer.from(`${sid}:${token}`).toString('base64')
    await this.postForm(endpoint, body, {
      Authorization: `Basic ${auth}`,
    })
  }

  private async sendEmailWebhook(payload: AlertPayload) {
    const headers: Record<string, string> = {}
    const token = this.config.ALERT_EMAIL_WEBHOOK_TOKEN.trim()
    if (token) {
      headers.Authorization = `Bearer ${token}`
    }

    const recipients = this.config.ALERT_EMAIL_TO.split(',').map((item) => item.trim()).filter(Boolean)

    await this.postJson(
      this.config.ALERT_EMAIL_WEBHOOK_URL,
      {
        subject: `[${severityPrefix[payload.severity]}] ${payload.title}`,
        text: this.toPlainText(payload),
        from: this.config.ALERT_EMAIL_FROM.trim() || undefined,
        to: recipients.length ? recipients : undefined,
        event: payload,
      },
      headers,
    )
  }

  private withWhatsappPrefix(value: string) {
    const trimmed = value.trim()
    if (!trimmed) {
      return trimmed
    }

    return trimmed.toLowerCase().startsWith('whatsapp:') ? trimmed : `whatsapp:${trimmed}`
  }

  private toPlainText(payload: AlertPayload) {
    const metadataRows = Object.entries(payload.metadata)
      .map(([key, value]) => `${key}: ${String(value)}`)
      .join('\n')

    return [
      `[${severityPrefix[payload.severity]}] ${payload.title}`,
      payload.message,
      `Source: ${payload.source}`,
      `Time: ${payload.occurredAt}`,
      metadataRows ? `Details:\n${metadataRows}` : '',
    ]
      .filter(Boolean)
      .join('\n')
  }

  private async postJson(url: string, body: unknown, extraHeaders: Record<string, string> = {}) {
    const controller = new AbortController()
    const timeout = setTimeout(() => controller.abort(), ALERT_HTTP_TIMEOUT_MS)

    try {
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          ...extraHeaders,
        },
        body: JSON.stringify(body),
        signal: controller.signal,
      })

      if (!response.ok) {
        const text = await response.text().catch(() => '')
        throw new Error(`HTTP ${response.status} ${response.statusText}${text ? `: ${text.slice(0, 300)}` : ''}`)
      }
    } finally {
      clearTimeout(timeout)
    }
  }

  private async postForm(url: string, body: URLSearchParams, extraHeaders: Record<string, string> = {}) {
    const controller = new AbortController()
    const timeout = setTimeout(() => controller.abort(), ALERT_HTTP_TIMEOUT_MS)

    try {
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
          ...extraHeaders,
        },
        body: body.toString(),
        signal: controller.signal,
      })

      if (!response.ok) {
        const text = await response.text().catch(() => '')
        throw new Error(`HTTP ${response.status} ${response.statusText}${text ? `: ${text.slice(0, 300)}` : ''}`)
      }
    } finally {
      clearTimeout(timeout)
    }
  }
}

File: Server/app/services/alert_service.py
from __future__ import annotations

import json
from datetime import datetime, timezone
from typing import Any

import httpx

from app.config import settings


class AlertService:
    def __init__(self) -> None:
        self._last_sent: dict[str, float] = {}

    async def notify(self, event: dict) -> None:
        if not settings.ALERTS_ENABLED_RESOLVED:
            return

        now = datetime.now(timezone.utc).timestamp() * 1000
        key = event.get("key", "unknown")
        last = self._last_sent.get(key, 0)
        if now - last < settings.ALERT_COOLDOWN_MS:
            return

        self._last_sent[key] = now
        payload = self._build_payload(event)

        tasks = []
        if settings.ALERT_SLACK_WEBHOOK_URL:
            tasks.append(self._send_slack(payload))
        if self._has_twilio():
            tasks.append(self._send_whatsapp(payload))
        if settings.ALERT_EMAIL_WEBHOOK_URL:
            tasks.append(self._send_email(payload))

        if not tasks:
            return

        import asyncio
        results = await asyncio.gather(*tasks, return_exceptions=True)
        failures = [r for r in results if isinstance(r, Exception)]
        if len(failures) == len(results):
            self._last_sent.pop(key, None)

    def _build_payload(self, event: dict) -> dict:
        return {
            "key": event.get("key"),
            "severity": event.get("severity", "info"),
            "title": (event.get("title") or "").strip(),
            "message": (event.get("message") or "").strip(),
            "metadata": event.get("metadata", {}),
            "occurred_at": datetime.now(timezone.utc).isoformat(),
            "source": "derton-finance-server",
        }

    def _has_twilio(self) -> bool:
        return bool(
            settings.ALERT_WHATSAPP_TWILIO_ACCOUNT_SID.strip()
            and settings.ALERT_WHATSAPP_TWILIO_AUTH_TOKEN.strip()
            and settings.ALERT_WHATSAPP_TWILIO_FROM.strip()
            and settings.ALERT_WHATSAPP_TWILIO_TO.strip()
        )

    async def _send_slack(self, payload: dict) -> None:
        text = f"[{payload['severity'].upper()}] {payload['title']}\n{payload['message']}"
        async with httpx.AsyncClient(timeout=10) as client:
            await client.post(
                settings.ALERT_SLACK_WEBHOOK_URL,
                json={"text": text},
            )

    async def _send_whatsapp(self, payload: dict) -> None:
        sid = settings.ALERT_WHATSAPP_TWILIO_ACCOUNT_SID.strip()
        token = settings.ALERT_WHATSAPP_TWILIO_AUTH_TOKEN.strip()
        body = f"[{payload['severity'].upper()}] {payload['title']}\n{payload['message']}"
        async with httpx.AsyncClient(timeout=10) as client:
            await client.post(
                f"https://api.twilio.com/2010-04-01/Accounts/{sid}/Messages.json",
                data={
                    "From": f"whatsapp:{settings.ALERT_WHATSAPP_TWILIO_FROM.strip()}",
                    "To": f"whatsapp:{settings.ALERT_WHATSAPP_TWILIO_TO.strip()}",
                    "Body": body,
                },
                auth=(sid, token),
            )

    async def _send_email(self, payload: dict) -> None:
        headers = {}
        token = settings.ALERT_EMAIL_WEBHOOK_TOKEN.strip()
        if token:
            headers["Authorization"] = f"Bearer {token}"

        async with httpx.AsyncClient(timeout=10) as client:
            await client.post(
                settings.ALERT_EMAIL_WEBHOOK_URL,
                json={
                    "subject": f"[{payload['severity'].upper()}] {payload['title']}",
                    "text": f"{payload['title']}\n{payload['message']}",
                    "event": payload,
                },
                headers=headers,
            )
